### PR TITLE
Fix QgsProxyProgressTask can get stuck if it's finalized before the task starts

### PR DESCRIPTION
**Pull request by nyalldawson**
Fixes a number of issues where tasks get "stuck" in the manager, and also should fix the long-standing timeout issues with the attribute table tests!

### CODE REVIEW

**Comment by nirvn** [comment]
Nice.

File: src/core/qgsproxyprogresstask.cpp [hunk 1]
```diff
@@ -25,15 +25,27 @@ QgsProxyProgressTask::QgsProxyProgressTask( const QString &description )
 
 void QgsProxyProgressTask::finalize( bool result )
 {
+  mNotFinishedMutex.lock();
```
Comment (elpaso) [inline src/core/qgsproxyprogresstask.cpp]: Can you use a locker here?

File: src/core/qgsproxyprogresstask.cpp [hunk 2]
```diff
@@ -25,15 +25,27 @@ QgsProxyProgressTask::QgsProxyProgressTask( const QString &description )
...
 {
   mNotFinishedMutex.lock();
-  mNotFinishedWaitCondition.wait( &mNotFinishedMutex );
-  mNotFinishedMutex.unlock();
+  if ( !mAlreadyFinished )
+  {
+    mNotFinishedWaitCondition.wait( &mNotFinishedMutex );
+    mNotFinishedMutex.unlock();
+  }
+  else
```
Comment (elpaso) [inline src/core/qgsproxyprogresstask.cpp]: Is this else necessary?

Comment (nyalldawson) [reply src/core/qgsproxyprogresstask.cpp]: I guess not :)

**Pull request merged**